// three/src/materials/nodes/MeshLambertNodeMaterial.js
import NodeMaterial from 'three/src/materials/nodes/NodeMaterial.js';
import BasicEnvironmentNode from 'three/src/nodes/lighting/BasicEnvironmentNode.js';
import PhongLightingModel from 'three/src/nodes/functions/PhongLightingModel.js';

import { MeshLambertMaterial } from 'three/src/materials/MeshLambertMaterial.js';

const _defaultValues = /*@__PURE__*/ new MeshLambertMaterial();

class MeshLambertNodeMaterial extends NodeMaterial {

	static get type() {

		return 'MeshLambertNodeMaterial';

	}

	constructor( parameters ) {

		super();

		this.isMeshLambertNodeMaterial = true;

		this.lights = true;

		this.setDefaultValues( _defaultValues );

		this.setValues( parameters );

	}

	setupEnvironment( builder ) {

		const envNode = super.setupEnvironment( builder );

		return envNode ? new BasicEnvironmentNode( envNode ) : null;

	}

	setupLightingModel( /*builder*/ ) {

		return new PhongLightingModel( false ); // ( specular ) -> force lambert

	}

}

export default MeshLambertNodeMaterial;
