// three/src/materials/nodes/MeshBasicNodeMaterial.js
import NodeMaterial from 'three/src/materials/nodes/NodeMaterial.js';
import { materialLightMap } from 'three/src/nodes/accessors/MaterialNode.js';
import BasicEnvironmentNode from 'three/src/nodes/lighting/BasicEnvironmentNode.js';
import BasicLightMapNode from 'three/src/nodes/lighting/BasicLightMapNode.js';
import BasicLightingModel from 'three/src/nodes/functions/BasicLightingModel.js';
import { normalView } from 'three/src/nodes/accessors/Normal.js';
import { diffuseColor } from 'three/src/nodes/core/PropertyNode.js';

import { MeshBasicMaterial } from 'three/src/materials/MeshBasicMaterial.js';

const _defaultValues = /*@__PURE__*/ new MeshBasicMaterial();

class MeshBasicNodeMaterial extends NodeMaterial {

	static get type() {

		return 'MeshBasicNodeMaterial';

	}

	constructor( parameters ) {

		super();

		this.isMeshBasicNodeMaterial = true;

		this.lights = true;

		this.setDefaultValues( _defaultValues );

		this.setValues( parameters );

	}

	setupNormal() {

		return normalView; // see #28839

	}

	setupEnvironment( builder ) {

		const envNode = super.setupEnvironment( builder );

		return envNode ? new BasicEnvironmentNode( envNode ) : null;

	}

	setupLightMap( builder ) {

		let node = null;

		if ( builder.material.lightMap ) {

			node = new BasicLightMapNode( materialLightMap );

		}

		return node;

	}

	setupOutgoingLight() {

		return diffuseColor.rgb;

	}

	setupLightingModel() {

		return new BasicLightingModel();

	}

}

export default MeshBasicNodeMaterial;
