// three/src/nodes/lighting/HemisphereLightNode.js
import AnalyticLightNode from 'three/src/nodes/lighting/AnalyticLightNode.js';
import { uniform } from 'three/src/nodes/core/UniformNode.js';
import { mix } from 'three/src/nodes/math/MathNode.js';
import { normalView } from 'three/src/nodes/accessors/Normal.js';
import { lightPosition } from 'three/src/nodes/accessors/Lights.js';
import { renderGroup } from 'three/src/nodes/core/UniformGroupNode.js';

import { Color } from 'three/src/math/Color.js';

class HemisphereLightNode extends AnalyticLightNode {

	static get type() {

		return 'HemisphereLightNode';

	}

	constructor( light = null ) {

		super( light );

		this.lightPositionNode = lightPosition( light );
		this.lightDirectionNode = this.lightPositionNode.normalize();

		this.groundColorNode = uniform( new Color() ).setGroup( renderGroup );

	}

	update( frame ) {

		const { light } = this;

		super.update( frame );

		this.lightPositionNode.object3d = light;

		this.groundColorNode.value.copy( light.groundColor ).multiplyScalar( light.intensity );

	}

	setup( builder ) {

		const { colorNode, groundColorNode, lightDirectionNode } = this;

		const dotNL = normalView.dot( lightDirectionNode );
		const hemiDiffuseWeight = dotNL.mul( 0.5 ).add( 0.5 );

		const irradiance = mix( groundColorNode, colorNode, hemiDiffuseWeight );

		builder.context.irradiance.addAssign( irradiance );

	}

}

export default HemisphereLightNode;
